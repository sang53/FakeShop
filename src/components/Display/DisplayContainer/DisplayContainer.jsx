import classes from "./DisplayContainer.module.css";
import PropTypes from "prop-types";
import { Link } from "react-router";

export default function DisplayContainer({ displayItems }) {
  const output = displayItems.map((obj) => {
    return (
      <Link to={obj.link} key={obj.id} className="product">
        <div className="product-img">
          <img src={obj.image} alt={obj.title} />
        </div>
        <div>
          <div>{obj.title}</div>
          {obj.price && <div>{obj.price}</div>}
        </div>
      </Link>
    );
  });
  return (
    <div className={"flex gap-1rem pad-1rem " + classes.displayContainer}>
      {output.length ? output : "Loading..."}
    </div>
  );
}

DisplayContainer.propTypes = {
  displayItems: PropTypes.array.isRequired,
};
