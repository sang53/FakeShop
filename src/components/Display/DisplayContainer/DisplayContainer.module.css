.displayContainer {
  flex-wrap: wrap;
  justify-content: center;
  background-color: var(--carousel-bgc);
}
