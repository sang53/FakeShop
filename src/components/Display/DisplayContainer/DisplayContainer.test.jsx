import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router";
import { describe, expect, test } from "vitest";
import DisplayContainer from "./DisplayContainer";
import { LocationDisplay } from "../../../test/utils/utils";

const testCategories = [
  {
    id: 1,
    image: "#",
    title: "Example Category",
    link: "/categories/1",
  },
  {
    id: 10,
    image: "#",
    title: "Example Category1",
    link: "/categories/10",
  },
];

const testProducts = testCategories.map((category) => {
  return { ...category, link: `/products/${category.id}`, price: 32.94 };
});

function renderDisplayContainer(value) {
  return render(
    <MemoryRouter initialEntries={["/"]}>
      <DisplayContainer displayItems={value} />
      <LocationDisplay />
    </MemoryRouter>
  );
}

test("shows loading screen on empty display", () => {
  const { container } = renderDisplayContainer([]);
  expect(() => screen.getAllByRole("link")).toThrowError();
  expect(container.querySelector("div")).toHaveTextContent(/loading/i);
});

describe("categories", () => {
  test("displays valid categories", () => {
    renderDisplayContainer(testCategories);
    expect(
      screen.getByRole("link", { name: /example category /i })
    ).toBeInTheDocument();
    expect(
      screen.getByRole("link", { name: /example category1/i })
    ).toBeInTheDocument();
  });
  test("category link", async () => {
    renderDisplayContainer(testCategories);
    await user.click(screen.getByRole("link", { name: /example category /i }));
    expect(screen.getByTestId("location-display")).toHaveTextContent(
      `/categories/${testCategories[0].id}`
    );
    await user.click(screen.getByRole("link", { name: /example category1/i }));
    expect(screen.getByTestId("location-display")).toHaveTextContent(
      `/categories/${testCategories[1].id}`
    );
  });
  test("categories display as expected", () => {
    const { container } = renderDisplayContainer(testCategories);
    expect(container).toMatchSnapshot();
  });
});

describe("products", () => {
  test("displays valid products", () => {
    renderDisplayContainer(testProducts);
    expect(
      screen.getByRole("link", { name: /example category /i })
    ).toBeInTheDocument();
    expect(
      screen.getByRole("link", { name: /example category1/i })
    ).toBeInTheDocument();
  });
  test("products link", async () => {
    renderDisplayContainer(testProducts);
    await user.click(screen.getByRole("link", { name: /example category /i }));
    expect(screen.getByTestId("location-display")).toHaveTextContent(
      `/products/${testProducts[0].id}`
    );
    await user.click(screen.getByRole("link", { name: /example category1 /i }));
    expect(screen.getByTestId("location-display")).toHaveTextContent(
      `/products/${testProducts[1].id}`
    );
  });
  test("products display as expected", () => {
    const { container } = renderDisplayContainer(testProducts);
    expect(container).toMatchSnapshot();
  });
});
