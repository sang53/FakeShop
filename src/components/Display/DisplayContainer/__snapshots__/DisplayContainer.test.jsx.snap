// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`categories > categories display as expected 1`] = `
<div>
  <div
    class="flex gap-1rem pad-1rem _displayContainer_acfe5e"
  >
    <a
      class="product"
      data-discover="true"
      href="/categories/1"
    >
      <div
        class="product-img"
      >
        <img
          alt="Example Category"
          src="#"
        />
      </div>
      <div>
        <div>
          Example Category
        </div>
      </div>
    </a>
    <a
      class="product"
      data-discover="true"
      href="/categories/10"
    >
      <div
        class="product-img"
      >
        <img
          alt="Example Category1"
          src="#"
        />
      </div>
      <div>
        <div>
          Example Category1
        </div>
      </div>
    </a>
  </div>
  <div
    data-testid="location-display"
  >
    /
  </div>
</div>
`;

exports[`products > products display as expected 1`] = `
<div>
  <div
    class="flex gap-1rem pad-1rem _displayContainer_acfe5e"
  >
    <a
      class="product"
      data-discover="true"
      href="/products/1"
    >
      <div
        class="product-img"
      >
        <img
          alt="Example Category"
          src="#"
        />
      </div>
      <div>
        <div>
          Example Category
        </div>
        <div>
          32.94
        </div>
      </div>
    </a>
    <a
      class="product"
      data-discover="true"
      href="/products/10"
    >
      <div
        class="product-img"
      >
        <img
          alt="Example Category1"
          src="#"
        />
      </div>
      <div>
        <div>
          Example Category1
        </div>
        <div>
          32.94
        </div>
      </div>
    </a>
  </div>
  <div
    data-testid="location-display"
  >
    /
  </div>
</div>
`;
